File: circles.java
// Stephen Fulwider
//	Driving in Circles - 2010 High School Online Programming Contest

import java.io.File;
import java.util.Arrays;
import java.util.Scanner;


public class circles
{

	public static void main(String[] args) throws Exception
	{
		// I like to instantiate my classes this way so that I don't have to use static variables or methods
		new circles();
	}
	
	final double RAD=Math.PI/180; // for angle->radian conversion
	final double oo=1e50; // a big value for infinity
	
	// circle radii
	int Cn;
	int[] C=new int[10];
	
	// number of roads
	int Rn;
	
	// points on the map - at most 42 since 20*2+2=42
	int Pn;
	point[] P=new point[42];
	
	// distance matrix
	double[][] D=new double[42][42];
	
	circles() throws Exception
	{
		Scanner in = new Scanner(System.in);
		for (int T=in.nextInt(); T-->0; )
		{
			// read input
			Cn=in.nextInt();
			Rn=in.nextInt();
			Pn=0;
			
			for (int i=0; i<Cn; ++i)
				C[i]=in.nextInt();
			
			// clear out distance matrix
			for (double[] d : D) Arrays.fill(d, oo);
			
			// read in each road
			for (int i=0; i<Rn; ++i)
			{
				// get circle and angle
				int d=in.nextInt();
				int a=in.nextInt();
				
				// get from and to circle
				int f=d;
				int t=d+1;
				--f; --t; // make 0 based
				
				// put point on from and to circles
				P[Pn++]=new point(a,f);
				P[Pn++]=new point(a,t);
				
				// put this road in the distance matrix
				D[Pn-1][Pn-2]=D[Pn-2][Pn-1]=Math.abs(C[f]-C[t]);
			}
			
			// get start circle and angle
			int sc=in.nextInt()-1;
			int sa=in.nextInt();
			
			P[Pn++]=new point(sa,sc);
			
			// get finish circle and angle
			int fc=in.nextInt()-1;
			int fa=in.nextInt();

			P[Pn++]=new point(fa,fc);
			
			// get distances between all points on same circle
			for (int i=0; i<Pn; ++i)
				for (int j=i+1; j<Pn; ++j)
					if (P[i].c==P[j].c)
						D[i][j]=D[j][i]=cirDist(i,j);
			
			// run floyds algorithm to get shortest point between all points
			//	note: there are more efficient algorithms than this since
			//	we are only interested in shortest distance from a single
			//	point, but since the graph is so small I just use the
			//	simplest graph algorithm to code
			for (int k=0; k<Pn; ++k)
				for (int i=0; i<Pn; ++i)
					for (int j=0; j<Pn; ++j)
						D[i][j]=Math.min(D[i][j], D[i][k]+D[k][j]);
			
			// print the distance between the start and finish point, making sure to add a small amount for tolerance
			System.out.printf("%.2f%n", D[Pn-2][Pn-1]+1e-11);
		}
	}
	
	// arc length = R*theta, theta = angle between points i and j
	double cirDist(int i, int j)
	{		
		// theta is just the minimum angle between the two points, converted to radians
		double theta=Math.min(Math.abs(P[i].a-P[j].a), 360-Math.abs(P[i].a-P[j].a))*RAD;
		return C[P[i].c]*theta;
	}
	
	// helper class to hold an angle and which circle it's on
	class point
	{
		int a;
		int c;
		
		point(int a, int c)
		{
			this.a=a;
			this.c=c;
		}
	}

}

File: circles.cpp


#include <stdio.h>
#include <math.h>

struct vert{
   double angle;
   int index;
};

int absI(int i)
{
   if (i < 0) return -1 * i;
   return i;
}

double minD(double a, double b)
{
   if (a < b) return a;
   return b;
}

int main()
{
   // The graph should never have more than 42 nodes 
   double graph [42][42];
   struct vert verts[11][20];
   int nVert [11];
   int radii[11];

   // Lets use a nice infiniti value and PI value
   double oo = 987654321.0;
   double PI = 3.141592653589793;

   int t;

   FILE * in = stdin;

   // Go through each test case
   fscanf(in, "%d", &t);
   for (int c=1; c<=t; c++)
   {
      int circles, roads;
      fscanf(in, "%d", &circles);
      fscanf(in, "%d", &roads);

      // Go through the entire graph and fill infiniti
      for (int i=0; i<50; i++)
         for (int j=0; j<50; j++)
            graph[i][j] = oo;

      // Go through each vertex list resetting the size
      for (int i=0; i<11; i++)
         nVert[i] = 0;

      // Read in each of the radii values
      for (int i=1; i<=circles; i++)
      {
         int temp;
         fscanf(in, "%d", &temp);
         radii[i] = temp;
      }

      // Read in each vertex
      int numVerts=0;
      for (int r=0; r<roads; r++)
      {
         int level;
         int angle;
         fscanf(in, "%d", &level);
         fscanf(in, "%d", &angle);

         // When saving nodes note we keep an index
         // we assign to each end of a straight road
         // We use these verticies to form a graph
         // of the entire map of ucf

         // Save the first new node we have
         int loc = nVert[level];
         verts[level][loc].angle = angle;
         verts[level][loc].index = numVerts;
         nVert[level]++;
         numVerts++;

         // Save the second new node we have
         level++;
         loc = nVert[level];
         verts[level][loc].angle = angle;
         verts[level][loc].index = numVerts;
         nVert[level]++;
         numVerts++;
      
         // Get the distance we need to travel, the distance for side
         // roads is just the radius of the larger circle minus the
         // radius of the smaller circle
         double dist = radii[level] - radii[level-1];

         // Connect the two indecies remember side roads go both ways
         graph[numVerts-1][numVerts-2] = dist;
         graph[numVerts-2][numVerts-1] = dist;
      }

      // Add the nodes for start and finish
      int sc, sa, si, fc, fa, fi;
      fscanf(in, "%d %d %d %d", &sc, &sa, &fc, &fa);

      // Add the starting node
      int loc = nVert[sc];
      verts[sc][loc].angle = sa;
      si = verts[sc][loc].index = numVerts;
      nVert[sc]++;
      numVerts++;
      
      // Add the finishing node
      loc = nVert[fc];
      verts[fc][loc].angle = fa;
      fi = verts[fc][loc].index = numVerts;
      nVert[fc]++;
      numVerts++;

      // Add the level connections for each level
      // This code connects each node connections with
      // all other connects on each level!
      for (int i=1; i<=circles; i++)
      {
         for (int j=0; j<nVert[i]; j++)
         {
            for (int k=j+1; k<nVert[i]; k++)
            {
               // We calculate the angle first by taking the absolute value
               // of the angle between the two nodes we are given
               int finalAngle = absI(verts[i][j].angle - verts[i][k].angle);
               
               // Then we find which is quicker going left or going right
               // we can do this by getting the other angle with some trig
               // the other angle equals 360 - the first angle
               finalAngle = minD(finalAngle, 360 - finalAngle);
               
               // Now we use the arc length formula to find the distance
               // between the two nodes, to use arc length we first convert
               // the angles into radians, then we use the formula 
               // angle * radius to get the arc length of our path
               double ffAngle = PI*(finalAngle/180.0);
               double len = ffAngle * radii[i];
               
               // Now we set that length for our path, remember we can
               // go boths ways down the road so save both edges!
               int index1 = verts[i][j].index;
               int index2 = verts[i][k].index;
               graph[index1][index2] = len;
               graph[index2][index1] = len;
            }
         }
      }

      // Run Floyd's, its a shortest path algorithm thats easy to code,
      // look it up! it will find the shortest path between all points in 
      // the graph in O(n^3) time and is very handy in contest programming
      for (int k=0; k<numVerts; k++)
      {
         for (int i=0; i<numVerts; i++)
         {
            for (int j=0; j<numVerts; j++)
            {
               if (graph[i][j] > (graph[i][k] + graph[k][j]))
                  graph[i][j] = graph[i][k] + graph[k][j];
            }
         }
      }

      // Print our result which is the dist start to finish!
      // Round the output appropriately, so that we get correct results
      printf("%.2lf\n", graph[si][fi]);
   }

   return 0;
}
